curriculum/roadmap/llm.py

```
@@ -0,0 +1,75 @@
+                "responseMimeType": "application/json",
+                "temperature": self.temperature,
+            },
+        }).encode("utf-8")
+        url = ENDPOINT.format(model=self.model) + f"?key={self.api_key}"
+
+        last = None
+        for attempt in range(1, self.max_attempts + 1):
+            req = urllib.request.Request(
+                url, data=body, headers={"Content-Type": "application/json"})
+            try:
+                with urllib.request.urlopen(req, timeout=120) as resp:
```

Ruff (0.16.1)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)